src/ray/core_worker/core_worker_process.cc

```diff
@@ -113,11 +128,519 @@ std::shared_ptr<CoreWorker> CoreWorkerProcess::TryGetWorker() {
+  // check.
+  if (!options.assigned_worker_port.has_value()) {
+    // TODO(hjiang): In the next PR we will pass down port number and raylet id and use
+    // them directly. Then we need to rename `RegisterWorkerToRaylet` to
+    // `RegisterDriverToRaylet`.
+    Status raylet_client_status =
+        CoreWorker::RegisterWorkerToRaylet(*raylet_conn,
+                                            worker_context->GetWorkerID(),
+                                            options.worker_type,
+                                            worker_context->GetCurrentJobID(),
+                                            options.runtime_env_hash,
+                                            options.language,
+                                            options.node_ip_address,
+                                            options.serialized_job_config,
+                                            options.startup_token,
+                                            &local_raylet_id,
+                                            &assigned_port);
+    if (!raylet_client_status.ok()) {
+      // Avoid using FATAL log or RAY_CHECK here because they may create a core dump file.
+      RAY_LOG(ERROR).WithField(worker_id)
+          << "Failed to register worker to Raylet: " << raylet_client_status;
+      QuickExit();
+    }
+    RAY_CHECK_GE(assigned_port, 0);
+  }
+
+  auto local_raylet_client = std::make_shared<raylet::RayletClient>(
+      std::move(raylet_conn), std::move(grpc_client), worker_context->GetWorkerID());
+  auto connected = true;
+
+  auto core_worker_server =
+      std::make_unique<rpc::GrpcServer>(WorkerTypeString(options.worker_type),
+                                        assigned_port,
+                                        options.node_ip_address == "127.0.0.1");
+
+  // Set our own address.
+  RAY_CHECK(!local_raylet_id.IsNil());
+  auto rpc_address = rpc::Address();
+  rpc_address.set_ip_address(options.node_ip_address);
+  rpc_address.set_port(core_worker_server->GetPort());
+  rpc_address.set_raylet_id(local_raylet_id.Binary());
+  rpc_address.set_worker_id(worker_context->GetWorkerID().Binary());
+  RAY_LOG(INFO).WithField(worker_context->GetWorkerID()).WithField(local_raylet_id)
+      << "Initializing worker at address: " << rpc_address.ip_address() << ":"
+      << rpc_address.port();
+
+  auto gcs_client = std::make_shared<gcs::GcsClient>(options.gcs_options, worker_context->GetWorkerID());
+  RAY_CHECK_OK(gcs_client->Connect(io_service_));
+
+  if (RayConfig::instance().task_events_report_interval_ms() > 0) {
+    if (!task_event_buffer->Start().ok()) {
+      RAY_CHECK(!task_event_buffer->Enabled()) << "TaskEventBuffer should be disabled.";
+    }
+  }
+  std::shared_ptr<rpc::CoreWorkerClientPool> core_worker_client_pool =
+      std::make_shared<rpc::CoreWorkerClientPool>([this](const rpc::Address &addr) {
+        auto core_worker = GetCoreWorker();
```

Comment:
```suggestion
        auto &core_worker = GetCoreWorker();
```
for the rest too, just makes me worry less about copies when reading

Reply:
Had GetCoreWorker return a const ref instead and assign it to rvals only